thermald: 2.2 -> 2.3

Emantor opened the pull request:
###### Motivation for this change
Supports DPTF, which works in conjunction with the patches in kernel 5.8.

Also cleanup the derivation.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Emantor committed:
thermald: 2.2 -> 2.3

Also cleanup the derivation:
- remove unnecessary PKG_CONFIG_PATH export
- change gpl2 to gpl2Only
- add meta.changelog
- reformat inputs and sort alphabetically
last three suggested by @drewrisinger.

drewrisinger commented on pkgs/tools/system/thermald/default.nix:
```diff
@@ -1,24 +1,47 @@
-, pkgconfig, dbus, dbus-glib, libxml2, autoconf-archive }:
+{ stdenv,
+ fetchFromGitHub,
+ autoconf,
+ automake,
+ libtool ,
+ pkgconfig,
+ dbus,
+ dbus-glib,
+ libxml2,
+ autoconf-archive,
+ gtk-doc,
+ lzma,
+ upower,
+ libevdev }:
```
```suggestion
 , autoconf
 , automake
 , libtool
 pkgconfig,
 dbus,
 dbus-glib,
 libxml2,
 autoconf-archive,
 gtk-doc,
 lzma,
 upower,
 libevdev 
 }:
```
Commas typically go before package, and ``}:`` ends on its own line.
Try running ``nix-shell -p nixpkgs-fmt --run "path/to/this/file.nix"``

drewrisinger commented on pkgs/tools/system/thermald/default.nix:
```diff
@@ -1,24 +1,47 @@
+    libevdev
+  ];
 
   preConfigure = ''
```
```suggestion
  preConfigure = ''./autogen.sh";
```

drewrisinger reviewed (changes requested):
* [x] Formatting fixes (try ``nix-shell -p nixpkgs-fmt --run "nixpkgs-fmt /path/to/file.nix"``)
* [x] Add meta.changelog attribute
* [x] Change license to gpl2Only or other appropriate: https://discourse.nixos.org/t/lib-licenses-gpl3-co-are-now-deprecated/8206
* [x] Commits LGTM
* [x] Builds via ``nix-review``.

```
https://github.com/NixOS/nixpkgs/pull/97469
1 package built:
thermald
```

Emantor commented:
I ran:
```
$ nix-shell -p nixpkgs-fmt --run "nixpkgs-fmt pkgs/tools/system/thermald/default.nix"
these paths will be fetched (0.90 MiB download, 3.64 MiB unpacked):
  /nix/store/2yv66xci59v5pjl56zmhv5i1rrv1fizb-nixpkgs-fmt-1.0.0
  /nix/store/bgdqr36i8csfda3yfq5asbcnsdkwrklc-bash-interactive-4.4-p23-dev
copying path '/nix/store/bgdqr36i8csfda3yfq5asbcnsdkwrklc-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
copying path '/nix/store/2yv66xci59v5pjl56zmhv5i1rrv1fizb-nixpkgs-fmt-1.0.0' from 'https://cache.nixos.org'...
0 / 1 have been reformatted
```
I don't know why no reformatting was applied. Fixed up the sorting and all other issues manually.

drewrisinger commented:
> I don't know why no reformatting was applied. Fixed up the sorting and all other issues manually.

Strange. Honestly, it's been a while since I've used ``nixpkgs-fmt``, can't tell you what'd be causing that.

drewrisinger reviewed (approved):
* [x] Diff LGTM
* [x] Commits LGTM
* [x] Builds via ``nix-review``:

```
https://github.com/NixOS/nixpkgs/pull/97469
1 package built:
thermald
```

Ma27 commented:
According to the changelog version 2.3 requires Linux 5.8 or later, so we shouldn't merge it yet IMHO as we use 5.4 as default. Alternatively we could expose this as `thermald_2_3` in our package set.

Emantor commented:
> According to the changelog version 2.3 requires Linux 5.8 or later, so we shouldn't merge it yet IMHO as we use 5.4 as default. Alternatively we could expose this as `thermald_2_3` in our package set.

I think this is a mistake in the config file, thermald will work fine for older kernels, however the adaptive DPTF policy will not be supported:
```
$ uname -a
Linux dibooki 5.4.62 #1-NixOS SMP Thu Sep 3 09:27:11 UTC 2020 x86_64 GNU/Linux
$ systemctl status thermald
● thermald.service - Thermal Daemon Service
     Loaded: loaded (/nix/store/ax4g6pyj1psgp158xvj6qcvllcmdf4pq-unit-thermald.service/thermald.service; enabled; vend>
     Active: active (running) since Thu 2020-09-10 06:46:16 CEST; 26s ago
   Main PID: 1197 (thermald)
         IP: 0B in, 0B out
      Tasks: 2 (limit: 4915)
     Memory: 6.9M
        CPU: 46ms
     CGroup: /system.slice/thermald.service
             └─1197 /nix/store/6l8v8k8al8bdvwnw4wyfrikvriij0h7c-thermald-2.3/sbin/thermald --no-daemon --dbus-enable

Sep 10 06:46:16 dibooki systemd[1]: Started Thermal Daemon Service.
Sep 10 06:46:16 dibooki thermald[1197]: [1599713176][MSG]22 CPUID levels; family:model:stepping 0x6:4e:3 (6:78:3)
Sep 10 06:46:16 dibooki thermald[1197]: [1599713176][MSG]Polling mode is enabled: 4
Sep 10 06:46:16 dibooki thermald[1197]: [1599713176][MSG]sensor id 11 : No temp sysfs for reading raw temp
Sep 10 06:46:16 dibooki thermald[1197]: [1599713176][MSG]sensor id 11 : No temp sysfs for reading raw temp
Sep 10 06:46:16 dibooki thermald[1197]: [1599713176][MSG]sensor id 11 : No temp sysfs for reading raw temp
Sep 10 06:46:16 dibooki thermald[1197]: [1599713176][MSG]Using generated /var/run/thermald/thermal-conf.xml.auto
Sep 10 06:46:16 dibooki thermald[1197]: [1599713176][MSG]Using config file /var/run/thermald/thermal-conf.xml.auto
```

Emantor committed:
nixos/thermald: add adaptive mode

thermald >=2.3 supports the adaptive DPTF mode, in conjunction with
kernel 5.8.

Ma27 commented:
What's the reason to update the module in another PR?

Emantor commented:
Right, I've pushed the module change into this PR as well.

michaelpj commented on nixos/modules/services/hardware/thermald.nix:
```diff
@@ -23,6 +23,15 @@ in {
+        default = false;
+        description = ''
+          Whether to enable adaptive mode, only working on kernel versions greater than 5.8.
+          Thermald will detect this itself, safe to enable on kernel versions below 5.8.
```
Should this be enabled by default, if it's safe even if not supported by the kernel?

Emantor replied:
Two reasons for not enabling this by default:
1. Thermald doesn't enable it by default, so we shouldn't either to not surprise existing users with new behavior
2. There are some existing platforms where the adaptive doesn't work correctly, so enabling this should be a user decision, see https://github.com/intel/thermal_daemon/issues?q=is%3Aissue+is%3Aopen+adaptive, in case of issues the user can easily rollback this config option.

Emantor commented:
@Ma27 Is there something missing to push this forward?

jtojnar commented on pkgs/tools/system/thermald/default.nix:
```diff
@@ -1,26 +1,48 @@
-    export PKG_CONFIG_PATH="${dbus.dev}/lib/pkgconfig:$PKG_CONFIG_PATH"
-    ./autogen.sh
-  '';
+  preConfigure = "./autogen.sh";
```
This will make configure run twice:

https://github.com/intel/thermal_daemon/blob/a86a9fc1bbc3d44c452adf69a0e2a8df8933e165/autogen.sh#L15-L17

Emantor replied:
should be fixed by using `preConfigure = "NO_CONFIGURE=1 ./autogen.sh"; `

ajs124 replied:
Isn't that what we have `autoreconfHook` for?

jtojnar replied:
[`autoreconfHook`](https://github.com/NixOS/nixpkgs/blob/aaf1331bca72b60750bb69624a37f2aa81063e05/pkgs/build-support/setup-hooks/autoreconf.sh) just runs `autoreconf -vfi`, which might not be sufficient in some projects. (Does it run `gtkdocize` like `autogen.sh` does? Is that necessary?)

jtojnar commented on pkgs/tools/system/thermald/default.nix:
```diff
@@ -36,7 +58,8 @@ stdenv.mkDerivation rec {
     homepage = "https://01.org/linux-thermal-daemon";
-    license = licenses.gpl2;
+    changelog = "https://github.com/intel/thermal_daemon/blob/master/README.txt";
+    license = licenses.gpl2Only;
```
From the changelog:

```
Version 0.5
- License update to GPL v2 or later
```

Emantor replied:
Ah, I looked at the license, but did not check any of the source files. Fixed.

jtojnar commented on pkgs/tools/system/thermald/default.nix:
```diff
@@ -1,26 +1,48 @@
+    pkgconfig
+  ];
+  buildInputs = [
+    autoconf
+    autoconf-archive
+    automake
```
At least these and `libtool` should likely go to `nativeBuildInputs`.

Emantor replied:
done.

jtojnar commented on pkgs/tools/system/thermald/default.nix:
```diff
@@ -1,26 +1,48 @@
-
-  patchPhase = ''sed -e 's/upstartconfdir = \/etc\/init/upstartconfdir = $(out)\/etc\/init/' -i data/Makefile.am'';
+  nativeBuildInputs = [
+    gtk-doc
```
If you are building documentation, you need `devdoc` in `outputs`, otherwise it will be removed:

https://github.com/NixOS/nixpkgs/blob/c12cfa428c0ee2bf2d900c32d5f0096b80236ce8/pkgs/build-support/setup-hooks/multiple-outputs.sh#L41

Also I do not see how it could build docs without `docbook-xsl-nons`.

Emantor replied:
I am at a loss here, event with `docbook-xsl-nons` the build process does not process any files in the `docs` folder:
```
Making all in docs
make[2]: Entering directory '/build/source/docs'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/build/source/docs'
```

jtojnar replied:
Maybe it needs to be explicitly enabled with `--enable-gtk-doc` or something.

Emantor replied:
bingo, now I need to find out how to fix gtk-doc trying to download the docbook dtd.

jtojnar replied:
Add the DTD it is trying to build to `nativeBuildInputs` (e.g. `docbook_xml_dtd_45`), it will be picked up from `$XMLCATALOG` variable set by `findXMLCatalogs` setup hook.

Emantor replied:
Done.

jtojnar commented on pkgs/tools/system/thermald/default.nix:
```diff
@@ -33,10 +60,13 @@ stdenv.mkDerivation rec {
     cp ./data/thermal-conf.xml $out/etc/thermald/
   '';
 
+  outputDevdoc = "out";
```
Could you keep it in a separate output (`outputs = [ "out" "devdoc" ];`) and also respect https://discourse.nixos.org/t/document-attribute-ordering-in-package-expressions/4887?

Emantor replied:
I hope I got the ordering right.

jtojnar reviewed (approved):
LGTM. Thanks.

Emantor commented:
Thanks for the reviews, I learned a lot and collected bookmarks for guides to adhere to for future PRs :+1:

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/235
